Advance Management Of Data (#57)
Vaibhavpratapsingh22/Courses: 5-dbms/8-advance-management.md

# Chapter -7 Advance Management OF Data

In this chapter we shall discuss advance management of data to facilitate efficient and scalable analysis of large amounts of data through DDBMS and NoSQL.

## Distributed Database Management System(DDBMS)

> A distributed database is basically a database that is not limited to one system, it is spread over different sites, i.e, on multiple computers or over a network of computers. A distributed database system is located on various sited that don’t share physical components. This is required when a particular database needs to be accessed by various users globally. It needs to be managed such that for the users it looks like one single database. There are two types:

Reviewer: Nothing to be added here - Just one question out of curiosity - could you think of some DDBMS around you in your daily life?

Author: Yes, let us suppose there is a distributor of let's say aqua-guard. Then the distributor has n number of retailers. All the retailers have there own database(stock) and in case they get out of stock they directly tell the distributor and distributor gives them the stock. In case the distributor does not have stock he can arrange it from any other retailer. So like this, the distributor acts like a computer network and the retailers acts as sites.

Reviewer 2: Very nice example. @Vaibhavpratapsingh22 Keep it up.

4. **Database Recovery**: One of the common techniques used in DDBMS is replication of data across different sites. Replication of data automatically helps in data recovery if database in any site is damaged. Users can access data from other sites while the damaged site is being reconstructed. Thus, database failure may become almost inconspicuous to users.

5. **Support for Multiple Application Software**: Most organizations use a variety of application software each with its specific database support. DDBMS provides a uniform functionality for using the same data among different platforms.

### Scalability
Scalability refers to the capability of a system to handle a growing amount of work, or its potential to perform more total work in the same elapsed time when processing power is expanded to accommodate growth. A system is said to be scalable if it can increase its workload and throughput when additional resources are added. There are two broad categories for scaling database systems: vertical scaling and horizontal scaling.

> <p align="center"><img src="https://bitpanda-academy.imgix.net/nulld4cb2f71-953b-4308-8edc-63408bc35a1b/bitpanda-academy-intermediate-15-scaling-infographic.png?auto=compress%2Cformat&fit=min&fm=jpg&q=80&w=2100" height="" width=""></p>

Reviewer 2: Nice and simple Image. Good work there! I have one suggestion here, can we reduce the size a bit because the image has very little information to show but it's taking a lot of space.

1. **Vertical scaling**: Vertical scaling, also known as scaling up, is the process of adding resources, such as memory or more powerful CPUs to an existing server. Removing the memory or changing to a less powerful CPU is known as scaling down.

   Adding or replacing resources to a system typically results in performance gains, but realizing such gains often requires reconfiguration and downtime. Furthermore, there are limitations to the number of additional resources that can be applied to a single system, as well as to the software that uses the system.

> 2. **Horizontal scaling**: Horizontal scaling sometimes referred to as scaling out, is the process of adding more hardware to a system. This typically means adding nodes (new servers) to an existing system.

Reviewer: Can we do both at the same time? 
When do we need to scale up or scale-out? 
Is there something like scaling down as well - how is that approached? 
What happens to the company which is losing all of its customers and has a big database storage standing behind their walls?

Author: Should I add the answers in the chapter or here only or at both places.

Author: YES 
We need scale-up if our product is new and we are taking a chance to check if we increase the efficiency will there be more customer request for the product.
We need scale-out if we have a big company and lots of money. As scale-out means adding more machines. So we need manpower also.
I don't know about scaling down.
Company has many options as it has the user DATABASE
1. The company can check which product the max customers love and can launch the product with new features.
2. The company can send emails, SMS to the customers so that they feel connected to the company.
3. Happy customers equal good business so the company can focus on them and provide them with good offers and discounts.
and so on...

### Network Partitioning(NP)

> A network partition refers to network decomposition into relatively independent subnets for their separate optimization as well as network split due to the failure of network devices. In both cases the partition-tolerant behavior of subnets is expected. This means that even after the network is partitioned into multiple sub-systems, it still works correctly.

Reviewer: how is it different from scaling out? or backing up?

Author: In scaling-out, we add more machines to increase efficiency. Whereas in network partitioning we divide the nodes into different sections. So that we can have an easy check on there working, which is a better option than checking a complete scale-out.

### Partition Tolerance
Partition Tolerance is a guarantee that the system continues to operate despite arbitrary message loss or failure of part of the system. In other words, even if there is a network outage in the data center and some of the computers are unreachable, still the system continues to perform properly.

### Data  Fragmentation
Data Fragmentation is the task of dividing a table into a set of smaller tables. The subsets of the table are called fragments. Fragmentation can be of three types: horizontal, vertical, and hybrid (combination of horizontal and vertical). Horizontal fragmentation can further be classified into two techniques: primary horizontal fragmentation and derived horizontal fragmentation.

Fragmentation should be done in a way so that the original table can be reconstructed from the fragments. This is needed so that the original table can be reconstructed from the fragments whenever required. This requirement is called “re-constructiveness.” Fregmentation is divided into three categories as follow:

> 1. **Vertical Fragmentation**: In vertical fragmentation, the fields or columns of a table are grouped into fragments. In order to maintain re-constructiveness, each fragment should contain the primary key field(s) of the table. Vertical fragmentation can be used to enforce privacy of data.

Reviewer: How is this decided that site 1 will have only names and nothing else? and what happens if site 1 needs second and third column information as well - doesn't it create just extra overhead to send requests to multiple sites?

Author: This is just a simple example. 
Yeah that's true that if site 1 needs second and third column information it will create more efforts. But that is what it is made for means if site1 contains only names that mean there must be a huge demand of names only. So for that, we have made a separate site for name for the ease of the user.

<p align="center"><img src="https://i.stack.imgur.com/9OlRZ.jpg" height="" width="" ></p>

> **2. Horizontal Fragmentation**: Horizontal fragmentation groups the tuples of a table in accordance to values of one or more fields. Horizontal fragmentation should also confirm to the rule of reconstructiveness. Each horizontal fragment must have all columns of the original base table.

Reviewer: What if a new entry is needed to be created at site 1 - and the constraint on database is that email id must be unique for all the users? would the processor then run through all the sites and all the DBs and all the entries there?  Wouldnt that be just too much overhead and extra information processing!?

Author: There is no need to run through all the sites and all the DBs, we will just update the DB of site1. If in future we see that the site2 is demanding the same data repeatedly that we updated on site1. Then there we have 2 options;
1.  Either we always redirect site2 to site1 if the user wants the email-id data.
2. Or we copy the data from site1 in the DB of site2.

<p align="ceneter"><img src="https://i.stack.imgur.com/wXunq.jpg" height="" width=""></p>

> **3. Hybrid Fragmentation**: In hybrid fragmentation, a combination of horizontal and vertical fragmentation techniques are used. This is the most flexible fragmentation technique since it generates fragments with minimal extraneous information. However, reconstruction of the original table is often an expensive task.

Reviewer: Why there was a need to hybrid fragmentation? wasn't vertical and horizontal enough? what were they lacking in first place?

Author: In vertical fragmentation, we display the selected columns whereas, in horizontal, we display the selected rows of the table. Whereas in hybrid fragmentation we can combine both horizontal and vertical together. 
Hybrid fragmentation comes in the game when we need more detailed data.

2. Replication transparencies: It basically made user unaware of the existence of copies as we know that copies of data may be stored at multiple sites for better availability performance and reliability.
Fragmentation transparency:
It basically made user unaware about the existence of fragments it may be the vertical fragment or horizontal fragmentation.
3. Increased Reliability and availability: Reliability is basically defined as the probability that a system is running at a certain time whereas Availability is defined as the probability that the system is continuously available during a time interval. When the data and DBMS software are distributed over several sites one site may fail while other sites continue to operate and we are not able to only access the data that exist at the failed site and this basically leads to improvement in reliability and availability.
4. Easier Expansion: In a distributed environment expansion of the system in terms of adding more data, increasing database sizes, or adding more data, increasing database sizes or adding more processor is much easier.

> 5. Improved Performance: We can achieve inter-query and intr-query parallelism by executing multiple queries at different sites by breaking up a query into a number of sub-queries that basically executes in parallel which basically leads to improvement in performance.

Reviewer: Are there any disadvantages as well?

Author: YES 
1. If the number of sites increases it gets difficult to connect all the sites with one another.
2. Data security reduces as we are not able to check all the sites regularly.
3. Maintaining a proper standard i.e when the number of sites gets increased then it becomes very difficult to maintain the standard.

<p align="center"><img src="https://miro.medium.com/max/700/1*UnG2G7_h0kqI9IHtnUk3qg.png" height="" width=""></p>

3. **High Availability**: This condition states that every request gets a response on success/failure. Achieving availability in a distributed system requires that the system remains operational 100% of the time. Every client gets a response, regardless of the state of any individual node in the system. This metric is trivial to measure: either you can submit read/write commands, or you cannot. Hence, the databases are time independent as the nodes need to be available online at all times. This means that, unlike the previous example, we do not know if “Pikachu” or “Bulbasaur” was added first. The output could be either one. Hence why, high availability isn’t feasible when analyzing streaming data at high frequency.

<p align= "center"><img src="https://miro.medium.com/max/700/1*ABrjUrZAY6V1hEkFPYvC7A.png" height="" width=""></p>

> ## Types of NoSQL Systems

Reviewer 2: I really liked this section It was very easy to read and I was connected to it the whole time.

**Examples**: Amazon SimpleDB, CouchDB, MongoDB, Riak, Lotus Notes are popular Document originated DBMS systems.

### 3. Column-based
Column-oriented databases work on columns and are based on BigTable paper by Google. Every column is treated separately. The values of single column databases are stored contiguously.
They deliver high performance on aggregation queries like SUM, COUNT, AVG, MIN, etc. as the data is readily available in a column.

**Examples**: Column-based NoSQL databases are widely used to manage data warehouses, business intelligence, CRM, Library card catalogs, HBase, Cassandra, HBase, Hypertable are examples of a column-based database.

> ### 4. Graph-Based

Reviewer: Very well described!

A graph type database stores entities as well the relations amongst those entities. The entity is stored as a node with the relationship as edges. An edge gives a relationship between nodes. Every node and edge has a unique identifier. Compared to a relational database where tables are loosely connected, a Graph database is a multi-relational in nature. Traversing relationships as fast as they are already captured into the DB, and there is no need to calculate them.
Graph base databases mostly used for social networks, logistics, spatial data.

**Examples**: Neo4J, Infinite Graph, OrientDB, FlockDB are some popular graph-based databases.

## Summary

> The chapter contains Distributed Database Management System(DDBMS), advantaged of DDBMS, NoSQL databases and types of NoSQL systems .

Reviewer 2: Here, Summary could be improved, right now it looks like README.md file summary

## Summary

> So at the end, we now about DDBMS i.e distributed database management system, its types and advantages. Then we learned about fragmentation, its types.Moving on we discussed network partitioning and partition tolerance. At last we learned about NoSQL, its types with some examples.

Reviewer 2: ```suggestion
So now we know all 

So in the end, now we know what is DDBMS, Fragmentation, its types, and its advantages. We also discussed network partitioning and partition tolerance, and Lastly, we learned about NoSQL, its types with some examples.
```
